tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SavedModelBundle.java

```diff
@@ -178,33 +177,39 @@ public Exporter withFunction(ConcreteFunction function) {
      * @throws IOException if saved model or variable state cannot be written on disk
      */
     public void export() throws IOException {
-      if (functions.isEmpty() || session == null) {
+      if (functions.isEmpty()) {
         throw new IllegalStateException("Model should contain at least one valid function");
       }
-      Graph graph = session.graph();
-
-      // It is imperative to retrieve the graphDef after the saverDef, as the former might add
-      // new ops to the graph for saving and restoring the variables.
-      SaverDef saverDef = graph.saverDef();
-
-      MetaGraphDef.Builder metaGraphDef = metaGraphDefBuilder
-          .setSaverDef(saverDef)
-          .setGraphDef(graph.toGraphDef())
-          .setMetaInfoDef(MetaInfoDef.newBuilder().addAllTags(Arrays.asList(tags)));
-      functions.forEach((k, f) -> metaGraphDef.putSignatureDef(k, f.signature().asSignatureDef()));
-
-      // Make sure saved model directories exist
-      Path variableDir = Paths.get(exportDir, "variables");
-      variableDir.toFile().mkdirs();
-
-      // Save the variables state
-      session.save(variableDir.resolve("variables").toString());
-
-      // Save the graph
-      SavedModel savedModelDef = SavedModel.newBuilder().addMetaGraphs(metaGraphDef).build();
-      try (OutputStream file =
-          new FileOutputStream(Paths.get(exportDir, "saved_model.pb").toString())) {
-        savedModelDef.writeTo(file);
+      try (Graph graph = new Graph();
+          Session session = new Session(graph)) {
+
+        functions.values().forEach(graph::attachFunction);
+
+        session.runInit();
```

Comment:
`ConcreteFunction` doesn't store any variable values, so this is proper.  The higher level function api stores them in the function's init scope, which we'll have to handle here, but that's independent.

I'm thinking of updating the Exporter to work with a single session and multiple signatures, sort of like the reverse of `SavedFunction`, thoughts?  @karllessard?  ConcreteFunction isn't really what you want to export to export a model any more.

Reply:
I'm a bit confused about the naming of these different variant of a function. I think the `SavedFunction` and `ConcreteFunction` can share a common interface to allow you to pass it around independently of their nature. They both share similar endpoints, like `signature(), call(Map<String, Tensor>), call(Tensor)...`.

What I understand of a "concrete function" for TF is that it is an executable of TF operations in graph or eager mode which inputs/outputs is defined by a typed signature. Both `SavedFunction` and `ConcreteFunction` seems to comply with this status so it's a bit confusing. Especially that `SavedFunction` is already something else in the [proto definitions](https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/proto/framework/SavedFunction.java) imported from the core lib. In this case the `SavedFunction` is a container for one or more `ConcreteFunction`.

So, just brainstorming here:

Maybe that common interface can be the one named `ConcreteFunction`. Then the `SavedFunction` would either implement that interface or return an object that implements it, to mimic the proto paradigm (there is also a `SavedConcreteFunction` proto).

Then then `ConcreteFunction` in this PR could be renamed to something else and implements the new interface. I'm not too sure how to name it though... the main difference is that it can be attached to a specific execution environment (graph or eager) and is not self-contained. `OperableFunction`? `SubFunction`? `FunctionNode`? I really don't know, any better ideas please?

Reply:
> I'm a bit confused about the naming of these different variant of a function. I think the `SavedFunction` and `ConcreteFunction` can share a common interface to allow you to pass it around independently of their nature. They both share similar endpoints, like `signature(), call(Map<String, Tensor>), call(Tensor)...`.
> 
> What I understand of a "concrete function" for TF is that it is an executable of TF operations in graph or eager mode which inputs/outputs is defined by a typed signature. Both `SavedFunction` and `ConcreteFunction` seems to comply with this status so it's a bit confusing. Especially that `SavedFunction` is already something else in the [proto definitions](https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/proto/framework/SavedFunction.java) imported from the core lib. In this case the `SavedFunction` is a container for one or more `ConcreteFunction`.

The difference here is that `SavedFunction` isn't really a function at all, it's just feed/fetch coordinates for a `Session`.  You can't execute it as part of a graph or eager session, you can't differentiate it, you can't serialize it (well, you can, but state like variables is lost so not really), and you can't pass it to ops.  They only interface they share is having a `Signature` and a `call(Map<String, Tensor>)`.

Even though we could extract the `signature()` and `call(Map<String, Tensor>)` to an interface, I can't think of anywhere that interface would actually be used.  All of the uses cases for ConcreteFunction rely on the underlying `TF_Function` (and if you're taking about the ConcreteFunction creation methods, those will eventually be deprecated in favor of the high level function API).

I do agree it's a bit confusing though.  What if we detach the `SavedFunction` from SavedModelBundle, and from functions, and treat it as essentially a reusable `Session.Runner` configuration?  It would still be backed by `Signature` for use with saving, but we could move it to `Session` and name it something like `SessionInterface`.  That also helps with the Exporter: instead of exporting ConcreteFunctions (which doesn't really work anyways), you export a `Session` and a set of `SessionInterface`s (or eventually high level functions, once they are added).  That's essentially the same as it's doing now just named different.

Reply:
Hey @rnett , sorry for the delay but I would like to revive this conversation now. Just for the sake of backward compatibility, I think having a common interface for `signature()` and `call(Map<String, Tensor>)` would still be appropriate, I can see how a function from a saved model graph or a `TF_Function` could be passed interchangeably to an inference execution pipeline, for example. Plus, that would be more backward compatible with our current solution.

I like this idea of abstraction because in both cases, users just want to feed some inputs to operations that generates a bunch of outputs. They don't need to care if these operations are part of a subfunction or a main graph.

Reply:
Yeah, I can do that.  Do you have any naming preferences?  I would like to keep ConcreteFunction for the native backed function to match Python, maybe something like `CallableFunction`?  Also, what are your thoughts on the "session config" session function instead of `SavedFunction`?

Reply:
Ok for `CallableFunction`. And what about `SessionFunction` for the saved model one? Since it will implement `CallableFunction`, could make sense to keep naming it as a function?

Reply:
Alright, this should be fixed.  I still need to do a review of my own to check everything, but you should be good to do a high-level review.